### ISSUE TITLE
[bug] OneToOne inverse relationship not made

### ISSUE THREAD

**Issue by lookfirst** [mikro-orm/mikro-orm#151]
Looking at the database, the issue seems to be that dog.person isn't populated, but person.dog is.

https://github.com/lookfirst/mikro-example/blob/master/test/dog.spec.ts#L51

```
[query] begin
[query] insert into `Dog` default values [took 1 ms]
[query] insert into `Person` (`dog`, `email`, `id`) values (?, ?, ?) [took 1 ms]
[query] commit
[query] select `e0`.* from `Dog` as `e0` where `e0`.`id` = ? limit ? [took 1 ms]

TypeError: Cannot read property 'init' of null
    at Context.<anonymous> (test/dog.spec.ts:61:42)
```

**Comment by lookfirst** [comment]
I think also having control over the inverse would be good too. I don't need to create the dog.person column.

**Comment by B4nan** [comment]
You have the association defined badly. 

- owning side is the one with `inversedBy` (it points to the inversed side, but it is the owning side)
- inverse side has to point to owner via `mappedBy` (it is inverse side, as opposed to inverse**d** side)

**Comment by B4nan** [comment]
```typescript
@Entity()
export class Person {
	@OneToOne({ inversedBy: 'person' })
	dog: Dog;
}

@Entity()
export class Dog {
	@OneToOne({ mappedBy: 'dog' })
	person!: Person;
}
```

**Comment by lookfirst** [comment]
Ah, I see now.

Just as an exercise, with the above though, how can I load `dog.person`?

```
		const dog1 = await orm.em.findOne(Dog, {id: dog.id});
		assert.strictEqual((await dog1!.person.init()).id, person.id);
```

Still says init() is null.

**Comment by B4nan** [comment]
There is validation for this, but it was not triggered as you didn't model it as bidirectional (no `mappedBy` on either side). 

Looks like we are missing some docs here, I thought I had it somewhere... Basically what is in the collections section applies (about bi-directional relations): https://mikro-orm.io/collections/

**Comment by lookfirst** [comment]
Adding entity: doesn't fix it.

**Comment by lookfirst** [comment]
```
	@OneToOne({ entity: () => Dog, inversedBy: 'person' })
	dog: Dog;

	@OneToOne({entity: () => Person, mappedBy: 'dog' })
	person!: Person;
```

still init() is null and column isn't created.

**Comment by lookfirst** [comment]
This one is a pretty major one for me right now. It just isn't creating the related column.

**Comment by lookfirst** [comment]
Found a workaround for this issue. It makes sense that it would work this way, I just wish it was cleaner.

https://github.com/lookfirst/mikro-example/blob/master/test/dog.spec.ts#L51

```
	it('creates a dog with an owner', async () => {
		const person = new Person({id: 'person:1', email: 'foo@bar.com'});
		const dog = new Dog();
		person.dog = dog;
		dog.person = person;

		await orm.em.persistAndFlush(person);

		orm.em.clear();

		const dog1 = await orm.em.findOne(Dog, {id: dog.id});
		assert.strictEqual((await dog1!.person!.init()).id, person.id);
	});
```

```
https://github.com/lookfirst/mikro-example/blob/master/src/db/model/Dog.ts#L13

export class Dog {
	@PrimaryKey()
	id!: number;

	@OneToOne({inversedBy: 'dog', mappedBy: 'dog'})
	person!: Person;
}

https://github.com/lookfirst/mikro-example/blob/master/src/db/model/Person.ts#L12

@Entity()
export class Person {
	@PrimaryKey()
	id: string;

	@OneToOne({inversedBy: 'person', mappedBy: 'person'})
	dog!: Dog;

	constructor(person: Partial<Person>) {
		this.id = person.id!;
	}
}

```

**Comment by B4nan** [comment]
Well that is something that should not be even allowed. As I said, there is pretty precise validation for setting up bidirectional mapping, it was just not used for 1:1 (only for m:1/1:m). I will look into this today and let you know. 

Obviously there are tests for this and they all pass, so it must be some weirdness in your setup (one idea - I am using `ts-node` or compiling via `tsc`, no babel - not sure what you are using to run the code but I know there are some problems with babel and typescript).

**Comment by lookfirst** [comment]
You can see it all in the example project. I'm using straight up tsc.

**Comment by B4nan** [comment]
So I got it working. 

First problem for me was that you used default export in your cli-config, which was not supported (now it is). Then some problems with paths (I used debug command to find out hehe), used default `baseDir` and changed the config to: 

```typescript
export default {
	entitiesDirs: ['./build/src/db/model'],
	entitiesDirsTs: ['./src/db/model'],
	debug: true,
	type: 'sqlite',
	dbName: './build/test.sqlite',
	warnWhenNoEntities: true, // this is default, so not necessary
	cache: { enabled: false },
	namingStrategy: EntityCaseNamingStrategy,
} as Options;
```

Then changed the 1:1 defs as I proposed earlier. God that `init` missing error finally. I did add logging of params so I can see what happens - it correctly created both entities:

```sql
[query] begin
[query] insert into `Dog` default values [took 1 ms]
[]
[query] insert into `Person` (`dog`, `email`, `id`) values (?, ?, ?) [took 0 ms]
[ 1, 'foo@bar.com', 'person:1' ]
[query] commit
[query] select `e0`.* from `Dog` as `e0` where `e0`.`id` = ? limit ? [took 1 ms]
[ 1, 1 ]
```

Your particular use case was not really supported, although I remember I was working on this... It worked either when you searched by the reference or when populating it:

```typescript
const dog1 = await orm.em.findOne(Dog, {id: dog.id}, ['person']); // ok, populates person entity as well
const dog2 = await orm.em.findOne(Dog, {person: person.id}); // ok, only finds person id
```

With `3.0.0-alpha.12` it will work even without that, joining the other table and getting the referenced PK automatically, so you will have a reference there (not initialized). 

Now the select query looks like this:

```sql
[query] select `e0`.*, `e1`.`id` as `person` from `Dog` as `e0` left join `Person` as `e1` on `e0`.`id` = `e1`.`dog` where `e0`.`id` = ? limit ? [took 0 ms]
```

**Comment by lookfirst** [comment]
v13 now... please take a look at the example project again with Dog and Person.

If I do not add `{owner: true}` to both sides of the relation, then the column is not created.

Running `yarn; yarn test` should work now out of the box.

I'm also not seeing the nice left join.

**Comment by B4nan** [comment]
Again, it works like a charm for me, here is all I had to change:

```diff
@@ -56,13 +56,12 @@
 	it('creates a dog with an owner', async () => {
 		const dog = new Dog();
 		const person = new Person({id: 'person:1', email: 'foo@bar.com', dog});
-		dog.person = person;
 
 		await orm.em.persistAndFlush(person);
 
 		orm.em.clear();
 
-		const dog1 = await orm.em.findOne(Dog, {id: dog.id}, ['person']);
+		const dog1 = await orm.em.findOne(Dog, {id: dog.id});
 		assert.strictEqual((await dog1!.person!.init()).id, person.id);
 	});
 });
```

I guess that is the problem we are having here, I never used that line as it was not needed, but looks like it is causing the error (although it should be fine to have it there). I will make sure this is working and add 1:1 metadata validation so you can define only 1 owner. 

> If I do not add {owner: true} to both sides of the relation, then the column is not created.

Just to make sure - I hope you are not expecting the column to be created on both sides in the database schema, that is obviously not how it works :] Only owning side has the FK column, that is why inverse side selects via the join query written above. 

> I'm also not seeing the nice left join.

That is because you defined both sides as owners. That join is there only for inverse sides - only they need it as they do not have the FK column.

**Comment by lookfirst** [comment]
Well... in this case, I'm fine with Person 'owning' the relationship to Dog. But what I'm trying to consider for bigger picture, is both use cases. OneToOne could be bidirectional.

> That is because you defined both sides as owners.

Not in the example. It is just owner on the Person side.

**Comment by B4nan** [comment]
In your example it already is bidirectional. Still, one side is always the owning one and that only has the FK.

**Comment by lookfirst** [comment]
Ok, got it. Thanks!

**Comment by lookfirst** [comment]
Hmm.... I have some weird case where the inverse side of a 1-1 is not being set on a `findOne`. I wrote a test for it, but it is passing! In my own production code though, it is not. I must be doing something stupid, but cannot figure it out. =( Both of these pass as expected in the example, but I have something similar that isn't setting person.dog...

https://github.com/lookfirst/mikro-example/blob/master/test/db/dog.spec.ts#L68

```
		const person1 = await orm.em.findOne(Person, person.id, ['dog']);
		expect(person1!.dog.id).toEqual(dog1!.id);
		const person1 = await orm.em.findOne(Person, person.id, true);
		expect(person1!.dog.id).toEqual(dog1!.id);
```

**Comment by lookfirst** [comment]
Ah, nevermind... I know what is going wrong... a prior query is pulling out the person. Had to add `{refresh: true}`.

**Comment by lookfirst** [comment]
Ugh, I'm still seeing a weird case where the inverse still isn't being set and I have to force a `.init()`, but I can't explain it and tests pass.

**Comment by B4nan** [comment]
> Ugh, I'm still seeing a weird case where the inverse still isn't being set and I have to force a .init(), but I can't explain it and tests pass.

Can I see the code (the one that is failing)? Ideally the whole handler up until the issue, plus some explanation of what are entity relationships.

I also spotted one edge case few days ago with 1:1 inverse side when implementing eager loading, but that was about not having the reference at all (the select query for inverse side was without the join clause that gets the owner id).

**Comment by B4nan** [comment]
This is the problem I came across - when you populate 1:1 inverse from owner, then the owner is not set on the inverse:

```typescript
const bar = FooBar2.create('fb'); // owner
bar.baz = new FooBaz2('fz'); // inverse
await orm.em.persistAndFlush(bar);
orm.em.clear();

const repo = orm.em.getRepository(FooBar2);
const a = await repo.findOne(bar.id, ['baz']); // load owner by id, populate inverse
expect(a!.baz!.isInitialized()).toBe(true); // ok, inverse is loaded too
expect(a!.baz!.bar.isInitialized()).toBe(true); // fail, owner is not set on inverse
```

**Comment by lookfirst** [comment]
Yes, exactly something like that.

**Comment by lookfirst** [comment]
I think it has to do something with querying `bar` first...

**Comment by B4nan** [comment]
Fine than, as I already have a fix for that one :]

**Comment by lookfirst** [comment]
Great! =)

**Comment by B4nan** [comment]
Will need to work out some failing tests though, will try to publish it later today.

**Comment by lookfirst** [comment]
No worries man! Appreciate all the hard work you are doing. If I wasn't so busy building this app, I'd be trying to dive in more for you. Hopefully in the future...

**Comment by B4nan** [comment]
Should be fixed in `3.0.0-alpha.21`.
